scripts/irods/test/test_misc.py

```diff
@@ -196,3 +196,43 @@ def test_server_removes_leftover_temporary_rulebase_pid_files_on_startup__issue_
             for fn in [pid_file_1, pid_file_2, pid_file_3]:
                 self.assertFalse(os.path.exists(fn))
 
+    @unittest.skipIf(test.settings.RUN_IN_TOPOLOGY, "skip for topology testing")
+    def test_server_respawns_processes__issue_4977(self):
+        import psutil
+        import time
+        from collections import defaultdict
+        main_server = None
+        sescoln = self.admin.session_collection
+        # Find the main server process
+        for proc in psutil.process_iter():
+            # It won't have any parents.
+            if proc.name() == 'irodsServer' \
+               and proc.parent().name() != 'irodsServer':
+                main_server = proc
+                # Kill the agent factory and the rule server
+                for child in proc.children(recursive=True):
+                    child.kill()
+                break
+        #It should run every 5 seconds but it's probably best to wait a lot longer just in case
+        time.sleep(15)
+        object_name = sescoln+"/hello"
+        self.assertTrue(len(main_server.children()) >= 2)
+        # If the API is up after this it seems it should be entirely working.
+        self.admin.assert_icommand(['itouch', object_name])
+        seen_process_names = defaultdict(int)
+        for i in main_server.children():
+            seen_process_names[i.name()] += 1
+        # There should only be a single delay server instance
+        self.assertTrue(seen_process_names['irodsReServer'] == 1)
+        attribute = "a1"
+        # Test the delay server
+        self.admin.assert_icommand(['irule', '-r',
+                                    "irods_rule_engine_plugin-irods_rule_language-instance",
+                                    "'delay(\"\") { msiModAVUMetadata(\"-d\",\"%s\",\"add\",\"a1\",\"v1\",\"u2\"); writeLine(\"serverLog\",\"Did it, %s\") }'"%(object_name,object_name),
```

Comment:
The first thing i see is that your call to `delay()` is missing the hints argument. The `delay()` function needs a target to run the body of the delay rule against. Try replacing line 231 with the following:
```bash
'delay("<INST_NAME>irods_rule_engine_plugin-irods_rule_language-instance</INST_NAME>") { msiModAVUMetadata("-d", "{0}", "add", "a1", "v1", "u2"); writeLine("serverLog", "XXX FROM TEST XXX"); }'.format(object_name),
```
Another thing to keep in mind is that the delay server runs every 30 seconds by default. Normally, the tests adjusts the config file so that the delay server wakes up more often (like every second). For your test, I'd recommend going into `/etc/irods/server_config.json` and adjusting the delay server options under `advanced_settings`. The option you need to adjust is called `rule_engine_server_sleep_time_in_seconds`. You'll need to restart the server after updating the config file.

Here is the link explaining each option: https://docs.irods.org/4.2.9/system_overview/configuration/